### DISCUSSION TITLE
Question: How to add widgets / PDF form fields to rotated PDFs?

### DISCUSSION THREAD

**Discussion by MrVonkey** [pymupdf/PyMuPDF#1837] [Q&A] [answered]
Hi. Is there any possibility to add widgets, especially text input fields, to rotated PDFs taking into account the correct orientation?

Text input fields (widgets) are wrongly orientated if the rotation value of a PDF is not zero using the following snippet:

```
pdf = fitz.open(input_file)
page = pdf[0]
form_field_rect = fitz.Rect(510, 15, 560, 30)     
form_field_rect = form_field_rect * page.derotation_matrix   
widget = fitz.Widget()
widget.rect = form_field_rect
widget.field_type = fitz.PDF_WIDGET_TYPE_TEXT
widget.field_name = "text-test"
widget.field_value = "TEST"
page.add_widget(widget)
pdf.save(output_file, garbage=4, deflate=True, deflate_images=True, deflate_fonts=True)
```

Example with a rotation of 0 and 90:
![rotated-vs-not-rotated](https://user-images.githubusercontent.com/16926945/180836839-e8ce50b4-4b54-4631-9101-8f44afb54b23.png)

Setting the page rotation to 0 before adding the widget didn't changed the result. It hadn't any effect (tried with different values). 

Maybe related to #496.

Input files of the example:
[rotated.pdf](https://github.com/pymupdf/PyMuPDF/files/9183459/rotated.pdf)
[not-rotated.pdf](https://github.com/pymupdf/PyMuPDF/files/9183460/not-rotated.pdf)

Output files of the example:
[out-rotated.pdf](https://github.com/pymupdf/PyMuPDF/files/9183525/out-rotated.pdf)
[out-not-rotated.pdf](https://github.com/pymupdf/PyMuPDF/files/9183526/out-not-rotated.pdf)

**Comment by JorjMcKie** [comment]
let us put this to the Discussions tab first ...

**Comment by JorjMcKie** [comment, marked as answer]
I regret to say that this is not supported.
There is an **_annotation_** flag however (fields in fact **_are_** annotations), that prevents rotation even when the page is rotated: the annot's top-left rectangle point will remain fixed.
So you can try this:
```python
import fitz
doc=fitz.open()
page=doc.new_page()
r=fitz.Rect(100,300,200,330)
w=fitz.Widget()
w.rect=r
w.field_type=fitz.PDF_WIDGET_TYPE_TEXT
w.field_value="Mary"
w.text_fontsize=11
w.field_name="textfield"
annot=page.add_widget(w)
annot.flags
4
annot.set_flags(4 | 2**4)
page.set_rotation(90)
doc.ez_save("x.pdf")
```
But different PDF viewers behave differently. I haven't any doing what you expect. SumatraPDF (Windows) displays it correctly, but does not support field updates:
![grafik](https://user-images.githubusercontent.com/8290722/180881755-d9491431-a698-450d-b142-6c13f2a99f1c.png)

PDF-XChange Viewer does it alright, Adobe Acrobat does not, Foxit does not, Nitro PDF does not, etc.
Best you try it out.

**Reply by MrVonkey** [reply]
Thank you for your quick reply and the information regarding the annotation flags. Unfortunately, I cannot rely on the choice of the PDF viewer 😉

So in my use case I now replace pages with a rotation > 0 by using show_pdf_page() to rotate the content, inserting it into a new page with rotation=0 and replacing the old rotated page with the new one.

Thanks for your work and the great support.
